src/tools/tests/manifest2proto-test.ts

```diff
@@ -182,6 +193,187 @@ describe('manifest2proto', () => {
     assert.deepStrictEqual(connections.map(hc => hc.direction), ['READS', 'WRITES', 'READS_WRITES']);
   });
 
+  it('encodes concrete particle instance types', async () => {
+    const manifest = await Manifest.parse(`
+      particle Abc in 'a/b/c.js'
+        a: reads X {a: Text}
+        b: reads Y {b: Number}
+        c: writes Z {c: Boolean}
+
+      recipe
+        a: use #tag1 #tag2
+        b: map 'by-id'
+        c: create @persistent
+        Abc
+          a: a
+          b: b
+          c: c
+    `);
+    const connections = (await toProtoAndBack(manifest)).recipes[0].particles[0].connections;
+
+    assert.deepStrictEqual(connections, [
+      {
+        handle: 'handle0',
+        name: 'a',
+        type: {entity: {schema: {
+          names: ['X'],
+          fields: {a: {primitive: 'TEXT'}},
+          hash: 'eb8597be8b72862d5580f567ab563cefe192508d',
+        }}}
+      },
+      {
+        handle: 'handle1',
+        name: 'b',
+        type: {entity: {schema: {
+          names: ['Y'],
+          fields: {b: {primitive: 'NUMBER'}},
+          hash: '19caf7b30005cfd9b03c7f96ae526dc49995105f',
+        }}},
+      },
+      {
+        handle: 'handle2',
+        name: 'c',
+        type: {entity: {schema: {
+          names: ['Z'],
+          fields: {c: {primitive: 'BOOLEAN'}},
+          hash: 'e1ad5776554cda0ee2aad4c45dc0afda2ffc56b9',
+        }}},
+      }
+    ]);
+  });
+
+
+  it('encodes variable particle instance types', async () => {
+    const manifest = await Manifest.parse(`
+      particle Writer1
+        first: writes [Foo {a: Text, c: Text}]
+        second: writes [Bar {a: Text, b: Text}]
+      particle Writer2
+        first: writes [Foo {a: Text, b: Text}]
+        second: writes [Bar {a: Text, c: Text}]
+      particle Reader1
+        first: reads [~f with Foo {a: Text, b: Text, c: Text}]
```

Comment:
Something is fishy here. The type of a handle is at most `[Foo {a: Text}]`, this particle should not typecheck.

Interestingly though, the encoded type is correct.

Reply:
This is the reason that I added the null check code -- I'll fix the issue here and not patch the null.